File: implementation/csharp/GameEventScript.SyntaxHighlighter/src/GameEventScriptAnsiRenderer.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Text;

namespace GameEventScript.SyntaxHighlighter;

/// <summary>Optional pure string renderer. Terminal detection, NO_COLOR, console output and diagnostics belong to the embedding.</summary>
public static class GameEventScriptAnsiRenderer
{
    /// <summary>The CLI's default 16-color foreground palette; null leaves text unchanged.</summary>
    public static int? DefaultColor(GameEventScriptSyntaxKind kind) => kind switch
    {
        GameEventScriptSyntaxKind.Comment => 90,
        GameEventScriptSyntaxKind.String => 32,
        GameEventScriptSyntaxKind.Number => 34,
        GameEventScriptSyntaxKind.Keyword => 35,
        GameEventScriptSyntaxKind.Constant or GameEventScriptSyntaxKind.Register => 33,
        GameEventScriptSyntaxKind.Builtin or GameEventScriptSyntaxKind.Message or GameEventScriptSyntaxKind.Type or GameEventScriptSyntaxKind.Tag
            or GameEventScriptSyntaxKind.Function or GameEventScriptSyntaxKind.Module or GameEventScriptSyntaxKind.Label => 36,
        _ => null
    };

    /// <summary>Renders UTF-16 spans using the default palette or a custom category-to-ANSI-color mapping. Source characters are preserved verbatim.</summary>
    /// <exception cref="ArgumentException">Spans overlap, exceed source bounds, split a surrogate pair, or a palette returns a color outside 30..37, 39, 90..97.</exception>
    /// <exception cref="ArgumentNullException">source or spans is null.</exception>
    public static string Render(string source, IReadOnlyList<GameEventScriptHighlightSpan> spans, Func<GameEventScriptSyntaxKind, int?>? color = null)
    {
        if (source is null) throw new ArgumentNullException(nameof(source));
        if (spans is null) throw new ArgumentNullException(nameof(spans));
        color ??= DefaultColor;
        var result = new StringBuilder();
        var position = 0;
        int? active = null;
        void Switch(int? next)
        {
            if (active == next) return;
            if (active is not null) result.Append("\u001b[0m");
            if (next is not null) result.Append("\u001b[").Append(next.Value).Append('m');
            active = next;
        }
        foreach (var span in spans)
        {
            if (span is null || span.Start < position || span.Length <= 0 || span.Start > source.Length - span.Length || Splits(source, span.Start) || Splits(source, span.Start + span.Length))
                throw new ArgumentException("Invalid UTF-16 spans.", nameof(spans));
            if (span.Start > position)
            {
                Switch(null);
                result.Append(source, position, span.Start - position);
            }
            var next = color(span.Kind);
            if (next is not null && next is not (>= 30 and <= 37) && next != 39 && next is not (>= 90 and <= 97)) throw new ArgumentException("Invalid ANSI foreground color.", nameof(color));
            Switch(next);
            result.Append(source, span.Start, span.Length);
            position = span.Start + span.Length;
        }
        Switch(null);
        return result.Append(source, position, source.Length - position).ToString();
    }

    private static bool Splits(string text, int index) => index > 0 && index < text.Length && char.IsHighSurrogate(text[index - 1]) && char.IsLowSurrogate(text[index]);
}

File: implementation/swift/GameEventScriptSyntaxHighlighter/Sources/GameEventScriptSyntaxHighlighter/GameEventScriptAnsiRenderer.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

/// Optional pure string renderer. Terminal detection, NO_COLOR, output and diagnostics belong to the embedding.
public enum GameEventScriptAnsiRenderer {
    /// The CLI's default 16-color foreground palette; nil leaves text unchanged.
    public static func defaultColor(_ kind: GameEventScriptSyntaxKind) -> Int? {
        switch kind {
        case .comment: 90
        case .string: 32
        case .number: 34
        case .keyword: 35
        case .constant, .register: 33
        case .builtin, .message, .type, .tag, .function, .module, .label: 36
        default: nil
        }
    }

    /// Renders UTF-16 spans with the default palette or a custom category-to-ANSI-color mapping, preserving source characters verbatim.
    /// - Throws: `invalidSpans` for overlapping/out-of-bounds spans or split Unicode scalars;
    ///   `invalidAnsiColor` when a palette returns a color outside 30...37, 39, 90...97.
    public static func render(_ source: String, spans: [GameEventScriptHighlightSpan], color: (GameEventScriptSyntaxKind) -> Int? = defaultColor) throws -> String {
        let units = Array(source.utf16)
        var output = ""
        var position = 0
        var active: Int?

        func switchColor(_ next: Int?) {
            guard active != next else { return }
            if active != nil { output += "\u{1b}[0m" }
            if let next { output += "\u{1b}[\(next)m" }
            active = next
        }

        func splits(_ index: Int) -> Bool {
            index > 0 && index < units.count && (0xd800...0xdbff).contains(units[index - 1]) && (0xdc00...0xdfff).contains(units[index])
        }

        for span in spans {
            guard span.start >= position, span.length > 0, span.start <= units.count - span.length, !splits(span.start), !splits(span.start + span.length) else { throw GameEventScriptHighlightingError.invalidSpans }
            if span.start > position {
                switchColor(nil)
                output += String(decoding: units[position..<span.start], as: UTF16.self)
            }
            let next = color(span.kind)
            if let next, !(30...37).contains(next) && next != 39 && !(90...97).contains(next) { throw GameEventScriptHighlightingError.invalidAnsiColor }
            switchColor(next)
            output += String(decoding: units[span.start..<(span.start + span.length)], as: UTF16.self)
            position = span.start + span.length
        }
        switchColor(nil)
        return output + String(decoding: units[position...], as: UTF16.self)
    }
}
